osquery/tables/system/darwin/extended_attributes.cpp

```diff
@@ -0,0 +1,223 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
+#include <sys/xattr.h>
+
+#include <boost/filesystem.hpp>
+#include <boost/property_tree/json_parser.hpp>
+
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+#include <osquery/filesystem.h>
+
+namespace pt = boost::property_tree;
+
+namespace osquery {
+namespace tables {
+
+// Structure of any OS X extended attribute
+struct XAttrAttribute {
+  std::string attribute_data;
+  int return_value;
+  int buffer_length;
+};
+
+// Foward declares of all parse functions
+void parseWhereFrom(QueryData &results,
+                    const XAttrAttribute &x_att,
+                    const std::string &path,
+                    const std::string &directory);
+void parseQuarantineFile(QueryData &results,
+                         const XAttrAttribute &x_att,
+                         const std::string &path,
+                         const std::string &directory);
+
+// Extended attribute name to the required parse function. If there is no parse
+// function, the data is returned as is.
+const std::map<std::string,
+               void (*)(QueryData &,
+                        const XAttrAttribute &,
+                        const std::string &,
+                        const std::string &)> xParseMap = {
+    {"com.apple.metadata:kMDItemWhereFroms", parseWhereFrom},
+    {"com.apple.quarantine", parseQuarantineFile},
+};
+
+// Handle any errors thrown by sys/xattr
+std::string handleError() {
+  switch (errno) {
+  case ENOATTR:
+    return "No such attribute";
+  case ENOTSUP:
+    return "No system support, or support disabled";
+  case ERANGE:
+    return "Size too small to hold attribute";
+  case EPERM:
+    return "The named attribute is not permitted for this type of object";
+  case EINVAL:
+    return "Name is invalid, or options has an unsupported bit set";
+  case EISDIR:
+    return "The path is not a regular file";
+  case ENOTDIR:
+    return "Part of the path was not a directory";
+  case ENAMETOOLONG:
+    return "The filename was too long";
+  case EACCES:
+    return "Insufficient permissions to read the requested file";
+  case ELOOP:
+    return "Too many symbolic links were encountered";
+  case EFAULT:
+    return "The path or name is invalid";
+  case EIO:
+    return "There was an I/O error while reading";
+  default:
+    return "No data";
+  }
+}
+
+// Pull the requested extended attribute from the path and return
+// the XAttrAttribute structure
+struct XAttrAttribute getAttribute(const std::string &path,
+                                   const std::string attribute) {
+  struct XAttrAttribute x_att;
+  x_att.buffer_length =
+      getxattr(path.c_str(), attribute.c_str(), NULL, 0, 0, 0);
```

Comment:
what's with the NULL and 0's, watch out for any null pointers that should be `nullptr`